### PR TITLE
Adding google_backup_dr_backup_plan_association resource and datasource

### PR DESCRIPTION
**Pull request by niharika-98**
**Release Note Template for Downstream PRs (will be copied)**

See [Write release notes](https://googlecloudplatform.github.io/magic-modules/contribute/release-notes/) for guidance.

```release-note:new-resource
`google_backup_dr_backup_plan_association` (beta)
```

```release-note:new-datasource
`google_backup_dr_backup_plan_association` (beta)
```

### CODE REVIEW

File: mmv1/third_party/terraform/services/backupdr/data_source_backup_dr_backup_plan_association_test.go.tmpl [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+	"strings"
+	"github.com/hashicorp/terraform-plugin-testing/helper/resource"
+	"github.com/hashicorp/terraform-provider-google/google/acctest"
+	"github.com/hashicorp/terraform-provider-google/google/tpgresource"
+	transport_tpg "github.com/hashicorp/terraform-provider-google/google/transport"
+	"github.com/hashicorp/terraform-plugin-testing/terraform"
+
+)
+
+func TestAccDataSourceGoogleBackupDRBackupPlanAssociation_basic(t *testing.T) {
```
Comment (SarahFrench) [inline mmv1/third_party/terraform/services/backupdr/data_source_backup_dr_backup_plan_association_test.go.tmpl:16]: [The test failure](https://github.com/GoogleCloudPlatform/magic-modules/pull/12159#issuecomment-2440624535) here is: 

```
        Error: Invalid resource type
        
          on terraform_plugin_test.tf line 55, in resource "google_backup_dr_backup_plan" "foo":
          55: resource "google_backup_dr_backup_plan" "foo" { 
        
        The provider hashicorp/google does not support resource type
        "google_backup_dr_backup_plan".
```

When I look at the `provider_mmv1_resources.go` file that is [generated in the google-beta provider from this PR](https://github.com/modular-magician/terraform-provider-google-beta/blob/auto-pr-12159/google-beta/provider/provider_mmv1_resources.go) I cannot see `google_backup_dr_backup_plan` listed there. Have you already had a PR accepted for adding that resource? If so you may need to pull those changes into your branch.

**Comment by niharika-98** [comment]
Hi @SarahFrench , Addressed all comments and acceptance tests are included and passing. We are looking at providing TF support for this resource in Nov 11 rollout as product is GA. Can you pls prioritize reviewing this.

**Comment by SarahFrench** [comment]
Hi @niharika-98, thanks for your email about this PR. I am on PTO today but I’ll ask our partner team in Google if anyone can take over as reviewer to meet your timeline.

File: mmv1/third_party/terraform/services/backupdr/data_source_backup_dr_backup_plan_association_test.go.tmpl [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    allow_missing = "true" 
+}
+
+
+resource "google_backup_dr_backup_plan" "foo" { 
+  location = "us-central1" 
+  backup_plan_id = "bp-test-%{random_suffix}"
+  resource_type= "compute.googleapis.com/Instance"
+  backup_vault = google_backup_dr_backup_vault.my-backup-vault.name
+ depends_on= [ google_backup_dr_backup_vault.my-backup-vault ]
```
Comment (zli82016) [inline mmv1/third_party/terraform/services/backupdr/data_source_backup_dr_backup_plan_association_test.go.tmpl:89]: weird indentation

File: mmv1/third_party/terraform/services/backupdr/data_source_backup_dr_backup_plan_association_test.go.tmpl [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+	standard_schedule {
+	  recurrence_type = "HOURLY"
+	   hourly_frequency = 6
+	    time_zone = "UTC"
+	     backup_window {
+		    start_hour_of_day = 0
+		    end_hour_of_day = 24
+      }
+    }
+	}
```
Comment (zli82016) [inline mmv1/third_party/terraform/services/backupdr/data_source_backup_dr_backup_plan_association_test.go.tmpl:102]: weird indentation

File: mmv1/third_party/terraform/services/backupdr/data_source_backup_dr_backup_plan_association_test.go.tmpl [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+}
+
+resource "google_backup_dr_backup_plan_association" "bpa" { 
+  location = "us-central1" 
+  backup_plan_association_id = "bpa-test-%{random_suffix}"
```
Comment (zli82016) [inline mmv1/third_party/terraform/services/backupdr/data_source_backup_dr_backup_plan_association_test.go.tmpl:107]:
```suggestion
  backup_plan_association_id = "tf-test-bpa-test-%{random_suffix}"
```

File: mmv1/third_party/terraform/services/backupdr/data_source_backup_dr_backup_plan_association_test.go.tmpl [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+  resource_type= "compute.googleapis.com/Instance"
+  backup_plan = google_backup_dr_backup_plan.foo.name
+  lifecycle {
+    ignore_changes = [resource]
+  }
+}
+
+data "google_backup_dr_backup_plan_association" "bpa-test" {
+  location =  "us-central1"
+  backup_plan_association_id="bpa-test-%{random_suffix}"
```
Comment (zli82016) [inline mmv1/third_party/terraform/services/backupdr/data_source_backup_dr_backup_plan_association_test.go.tmpl:118]:
```suggestion
  backup_plan_association_id="tf-test-bpa-test-%{random_suffix}"
```

File: mmv1/third_party/terraform/website/docs/d/backup_dr_backup_plan_association.html.markdown [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+~> **Warning:** This resource is in beta, and should be used with the terraform-provider-google-beta provider.
+See [Provider Versions](https://terraform.io/docs/providers/google/guides/provider_versions.html) for more details on beta resources.
+
+## Example Usage
+
+```hcl
+data "google_backup_dr_backup_plan_association" "my-backupplan-association" {
+  location =  "us-central1"
+  backup_plan_association_id="bpa-id"
+  }
```
Comment (zli82016) [inline mmv1/third_party/terraform/website/docs/d/backup_dr_backup_plan_association.html.markdown:20]: weird indentation

File: mmv1/third_party/terraform/services/backupdr/data_source_backup_dr_backup_plan_association_test.go.tmpl [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+)
+
+func TestAccDataSourceGoogleBackupDRBackupPlanAssociation_basic(t *testing.T) {
+	t.Parallel()
+	context := map[string]interface{}{
+		"random_suffix": acctest.RandString(t, 10),
+	}
+	acctest.VcrTest(t, resource.TestCase{
+		PreCheck:                 func() { acctest.AccTestPreCheck(t) },
+		ProtoV5ProviderFactories: acctest.ProtoV5ProviderFactories(t),
```
Comment (zli82016) [inline mmv1/third_party/terraform/services/backupdr/data_source_backup_dr_backup_plan_association_test.go.tmpl:17]:
```suggestion
		ProtoV5ProviderFactories: acctest.ProtoV5ProviderBetaFactories(t),
```

Comment (zli82016) [reply mmv1/third_party/terraform/services/backupdr/data_source_backup_dr_backup_plan_association_test.go.tmpl:17]: Besides this change, `provider = google-beta` is needed for each resource in the configuration.

File: mmv1/templates/terraform/examples/backup_dr_bpa.tf.tmpl [hunk 8]
```diff
@@ -0,0 +1,77 @@
+ resource "google_service_account" "mySA" {
+  provider = google-beta
+  account_id   = "{{index $.Vars "account_idaccount_idaccount_id"}}"
```
Comment (zli82016) [inline mmv1/templates/terraform/examples/backup_dr_bpa.tf.tmpl:3]:
```suggestion
  account_id   = "{{index $.Vars "account_id"}}"
```

**Review by zli82016 — CHANGES_REQUESTED** [review]
Fix failed test

**Review by SarahFrench — COMMENTED** [review]
Thanks for helping with review @zli82016!

@niharika-98 I can see that these open review comments are yet to be addressed:

- [Query about whether ignore_changes is needed](https://github.com/GoogleCloudPlatform/magic-modules/pull/12159/files#r1828134440)
- [Request to change the acceptance test to use the Beta provider factory](https://github.com/GoogleCloudPlatform/magic-modules/pull/12159/files#r1828147163)
- Various requests to update resource names in acceptance test configuration, to make those resources identifiable as deletable if the test cannot delete those resources successfully itself:
    1) https://github.com/GoogleCloudPlatform/magic-modules/pull/12159/files#r1828141638
     2) https://github.com/GoogleCloudPlatform/magic-modules/pull/12159/files#r1828141872
    3) https://github.com/GoogleCloudPlatform/magic-modules/pull/12159/files#r1828142105

**Comment by niharika-98** [comment]
Addressed comments on PR

When we pass compute instance, project/projectname/locations/location/instances/instance-name as it is returned in id field of compute. But api returns projects/projectid/locations/locationid/instances/instanceid after creation completion in response. Hence there is mismatch so ignoring it. Compute instance name is parameter which is input to create BPA resource

VCR-test is giving error. Can't find exact string but it is similar to "Changes we present even after applying step 1/1. Terraform forces replacement"


Other comments are all addressed.

**Comment by zli82016** [comment]
> When we pass compute instance, project/projectname/locations/location/instances/instance-name as it is returned in id field of compute. But api returns projects/projectid/locations/locationid/instances/instanceid after creation completion in response. Hence there is mismatch so ignoring it. Compute instance name is parameter which is input to create BPA resource

@niharika-98 , thanks for addressing the comments. Instead of adding `ignore_changes` to the tests, can you please add `ignore_read: true` to the `resource` field to address this issue?

File: mmv1/products/backupdr/BackupPlanAssociation.yaml [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+    description: |
+      The location for the backupplan association
+  - name: 'backup_plan_association_id'
+    type: String
+    required: true
+    url_param_only: true
+    description: |-
+      The id of backupplan association
+properties:
+  - name: 'resource'
```
Comment (zli82016) [inline mmv1/products/backupdr/BackupPlanAssociation.yaml:58]: Can you add `ignore_read: true` to address the perma-diff issue?

File: mmv1/third_party/terraform/services/backupdr/data_source_backup_dr_backup_plan_association_test.go.tmpl [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+resource "google_backup_dr_backup_plan_association" "bpa" { 
+  provider = google-beta
+  location = "us-central1" 
+  backup_plan_association_id = "tf-test-bpa-test-%{random_suffix}"
+  resource =   google_compute_instance.default.id
+  resource_type= "compute.googleapis.com/Instance"
+  backup_plan = google_backup_dr_backup_plan.foo.name
+  lifecycle {
+    ignore_changes = [resource]
+  }
```
Comment (zli82016) [inline mmv1/third_party/terraform/services/backupdr/data_source_backup_dr_backup_plan_association_test.go.tmpl:120]: `add ignore_read: true` is not needed here.

**Review by zli82016 — CHANGES_REQUESTED** [review]

**Review by zli82016 — APPROVED** [review]
LGTM

**Pull request merged**